### ISSUE TITLE
Collection expressions can perform covariant conversion of ReadOnlySpan in emit

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#71106]
The runtime doesn't provide a safe facility for covariant conversion of ReadOnlySpan<T>, but it is supported to perform such conversions. If the compiler performed such conversions, it would enable more optimizations in scenarios like the following:

```cs
Span<Derived> span1 = ..., span2 = ...;
ReadOnlySpan<Base> spanBase1 = [..span1]; // emit as conversion to ReadOnlySpan<Base> then 'ToArray()'
ReadOnlySpan<Base> spanBase2 = [..span1, ..span2]; // emit as conversion then 'CopyTo()' the backing collection of 'spanBase2'
```

To start with, we are only going to perform ToArray and CopyTo optimizations in the case where the spread operator and the target collection have the same element type.

**Comment by CyrusNajmabadi** [comment]
Yes.  This would be nice.

**Comment by stephentoub** [comment]
> The runtime doesn't provide a safe facility for covariant conversion of ReadOnlySpan

There is now a `ReadOnlySpan<TBase>.CastUp<TDerived>` API in .NET 9.

**Comment by RikkiGibson** [comment]
I would review a PR which implements this by using the CastUp API.

**Comment by RikkiGibson** [comment]
@DoctorKrolic in #80104 we did "covariance on the output side". It would be great if we could also do "covariance on the input side", performing covariant conversions on `.. array` where needed, or converting to ROS and calling CastUp for `.. span` and `.. readOnlySpan`, and enable us to do covariant spread optimizations when the target collection is mutable (and exact matching result array type is needed). Let me know if you have any interest.

**Comment by DoctorKrolic** [comment]
> Let me know if you have any interest.

Yes, I ended up with https://github.com/dotnet/roslyn/pull/80104 after exploring this issue first. You can assign me here, but I cannot guarantee the timeline when I will deliver it, but I hope it to bee soon.

**Comment by RikkiGibson** [comment]
For this part of it we might want to get the `[with()]` feature merged to main first to avoid conflicts. I'll ping on this issue when that happens.
